Use React.Profiler to send render timings to GA

bobsilverberg opened the pull request:
Fixes #1423 

This will send data on the performance of the render function for both `CodeView` and `DiffView` to GA for future analysis. There has been some discussion about how much highlighting affects the performance, so I built in a separate measure of just the highlighting piece for the `CodeView` component.

bobsilverberg commented:
@willdurand I assigned this to the team, and Mat won the lottery, but I would like to have your feedback on the patch as well, if you don't mind, as it's something different from what we've ever done before. Thanks!

bobsilverberg commented:
> This seems to work (*) and make sense, but I'm not familiar with any of this.
> 
> (*) Weirdly, even if I had deactivated tracking protection for this site (and had deactivated my adblocker as well) Firefox kept answering `"1"` for `navigator.noNotTrack` (and the pref was set to follow tracking protection, it wasn't set to `always`) so I had to comment the code that deactivates tracking when DNT is enabled to see this patch in action. Not sure what's going on there...

There is a config setting called `trackingEnabled` which needs to be set to `true` in order for tracking to work. This is legacy and I'm not sure why it's set up that way, but it is. I have a file called `config/local-development-amo.js`, in which I have the following code:

```
module.exports = {
    enableStrictMode: false,
    trackingEnabled: true,
    allowErrorSimulation: true,
};
```

which configures my dev environment. Perhaps if you turn on `trackingEnabled` you'll find it works for you too?

bobsilverberg commented:
> > This seems to work (*) and make sense, but I'm not familiar with any of this.
> > (*) Weirdly, even if I had deactivated tracking protection for this site (and had deactivated my adblocker as well) Firefox kept answering `"1"` for `navigator.noNotTrack` (and the pref was set to follow tracking protection, it wasn't set to `always`) so I had to comment the code that deactivates tracking when DNT is enabled to see this patch in action. Not sure what's going on there...
> 
> There is a config setting called `trackingEnabled` which needs to be set to `true` in order for tracking to work. This is legacy and I'm not sure why it's set up that way, but it is. I have a file called `config/local-development-amo.js`, in which I have the following code:
> 
> ```
> module.exports = {
>     enableStrictMode: false,
>     trackingEnabled: true,
>     allowErrorSimulation: true,
> };
> ```
> 
> which configures my dev environment. Perhaps if you turn on `trackingEnabled` you'll find it works for you too?

As @willdurand pointed out, that code, above is actually from addons-frontend, and not addons-code-manager, so everything I said was irrelevant. I'll look into @diox's issue a bit more.

willdurand commented:
> This will send data on the performance of the render function for both `CodeView` and `DiffView` to GA for future analysis. There has been some discussion about how much highlighting affects the performance, so I built in a separate measure of just the highlighting piece for the `CodeView` component.

The issue mentions InfluxDB and not GA, so I guess we won't be using InfluxDB/Grafana?

willdurand commented on src/components/CodeView/index.tsx:
```diff
@@ -21,13 +21,16 @@ import {
 import GlobalLinterMessages from '../GlobalLinterMessages';
 import SlowPageAlert from '../SlowPageAlert';
 
+const { Profiler } = React;
```
That's weird but why not. Can't you import it as usual? (or use `React.Profiler`)

bobsilverberg replied:
Sure, I can try the former and if that doesn't work use the latter. I cannot remember exactly why I needed to do this.

willdurand commented on src/components/CodeView/index.tsx:
```diff
@@ -81,11 +85,16 @@ type Props = PublicProps & DefaultProps & RouteComponentProps;
     enableCommenting: process.env.REACT_APP_ENABLE_COMMENTING === 'true',
   };
 
+  onRenderCallback = (id: string, phase: string, actualDuration: number) => {
```
```suggestion
  onRenderProfiler = (id: string, phase: string, actualDuration: number) => {
```

willdurand replied:
We should likely add a link to the docs: https://github.com/reactjs/rfcs/blob/master/text/0051-profiler.md above this method.

willdurand commented on src/components/CodeView/index.tsx:
```diff
@@ -197,12 +206,17 @@ export class CodeViewBase extends React.Component<Props> {
+                            <Profiler
+                              id="CodeView-Highlighting"
+                              onRender={this.onRenderCallback}
+                            >
```
So we are going to profile every single line of code?

How are we going to determine that the `Profiler` isn't slowing down the page more than without `Profiler`? I read that impact should be minimal here: https://gist.github.com/bvaughn/25e6233aeb1b4f0cdb8d8366e54a3977#gistcomment-2744384, but this part of our code is going to render a `Profiler` for every single LOC.

bobsilverberg replied:
Ah, I didn't realize we were highlighting the code one line at a time - I thought we were doing it to all the code at once. It's odd, because when I tested this locally, I only saw one call to send the data to GA for highlighting per render, not a call for every line of code, so I wonder what exactly is happening here? Maybe `onRender` is only getting called once? If so, what exactly is being measured?

It's fine to just remove this for now, as I think we're more concerned at the moment with measuring overall performance of the component and not trying to isolate highlighting performance. That will resolve this particular issue.

willdurand commented on src/components/DiffView/index.tsx:
```diff
@@ -41,11 +41,14 @@ import {
 import styles from './styles.module.scss';
 import 'react-diff-view/style/index.css';
 
+const { Profiler } = React;
```
Same question.

willdurand commented on src/components/DiffView/index.tsx:
```diff
@@ -131,13 +135,18 @@ export class DiffViewBase extends React.Component<Props> {
     viewType: 'unified',
   };
 
+  onRenderCallback = (id: string, phase: string, actualDuration: number) => {
```
```suggestion
  onRenderProfiler = (id: string, phase: string, actualDuration: number) => {
```

willdurand replied:
We should likely add a link to the docs: https://github.com/reactjs/rfcs/blob/master/text/0051-profiler.md above this method.

willdurand commented on src/utils.tsx:
```diff
@@ -270,3 +271,22 @@ export const codeShouldBeTrimmed = (
+    label: phase,
+    value: actualDuration,
+    variable: id,
+  });
```
How is it looking in GA? Can we graph these metrics? Or can we use aggregate functions at least (to compute percentiles, etc.)?

bobsilverberg replied:
We have data in GA now for all the fetch timings and I've taken a look at that and it does look like we can manipulate it in useful ways. You can take a look too, but you might need to ask for access to the code manager sites in GA.

bobsilverberg commented:
> > This will send data on the performance of the render function for both `CodeView` and `DiffView` to GA for future analysis. There has been some discussion about how much highlighting affects the performance, so I built in a separate measure of just the highlighting piece for the `CodeView` component.
> 
> The issue mentions InfluxDB and not GA, so I guess we won't be using InfluxDB/Grafana?

Correct. I had a discussion on Slack in #infraeng-influxdb and the result was a decision not to go that route and use GA instead. I will update the issue.

bobsilverberg commented:
Thanks for the review @willdurand. I have pushed a new commit that addresses your issues. I am going to run an errand this morning, but hope to be back before noon. I'd love to get this into today's tag if possible.
